storage/innobase/srv/srv0start.cc

```diff
@@ -465,25 +465,33 @@ create_log_files(
 	ut_a(fil_validate());
 	ut_a(log_space != NULL);
 
-      /* Once the redo log is set to be encrypted,
-        initialize encryption information. */
-       if (srv_redo_log_encrypt != REDO_LOG_ENCRYPT_OFF) {
-               if (!Encryption::check_keyring()) {
-                       ib::error()
-                               << "Redo log encryption is enabled,"
-                               << " but keyring plugin is not loaded.";
-
-                       return(DB_ERROR);
-               }
-
-               log_space->flags |= FSP_FLAGS_MASK_ENCRYPTION;
-               err = fil_set_encryption(log_space->id,
-                                        Encryption::AES,
-                                        NULL,
-                                        NULL);
-               ut_ad(err == DB_SUCCESS);
-       }
+	/* Once the redo log is set to be encrypted,
+	   initialize encryption information. */
+	if (srv_redo_log_encrypt != REDO_LOG_ENCRYPT_OFF) {
+		if (!Encryption::check_keyring()) {
+			ib::error() << "Redo log encryption is enabled,"
+				    << " but keyring plugin is not loaded.";
 
+			return (DB_ERROR);
+		}
+
+		Encryption::Type alg =
+		    srv_redo_log_encrypt == REDO_LOG_ENCRYPT_RK
+			? Encryption::KEYRING
+			: Encryption::AES;
+
+		log_space->flags |= FSP_FLAGS_MASK_ENCRYPTION;
+
+		redo_log_key *mkey =
+		    redo_log_key_mgr.generate_new_key_without_storing();
```

Comment:
Whats the idea of "non-stored" version? I mean the advantage of using this variant?

 It seems always the first element in vector is used for this. When will this be stored?

Reply:
During bootstrap with redo encryption, we have to encrypt the log before we can communicate with the keyring plugin. This is why we generate the first key directly and store later when the keyring is guaranteed to be available.

Reply:
What you wrote is a good example of a function comment for generate_new_key_without_storing(). 

If you are open to accepting suggestions on name, my suggestion is 

generate_key_for_bootstrap().

One more doubt, will this be used when redo files are recreated on startup? In that case we shouldn't use this function?

Reply:
It's called in `create_log_files()`, so we use it every time when:

* there is no existing redo log
* redo log encryption is turned on

I could add more conditions / paths, but it only would make the code more complex, the key is written to the keyring a bit later anyway.

Reply:
Ok, lets assume the recreate case. We first create the key assuming keyring is not accessible, Will the key change once we know the keyring is available? When and how it will use a key generated by using keyring plugin?  (i.e my_key_generate())?

Reply:
Change: no, we upload this key to the keyring, and keep using it.

Key generated by keyring: every other case generates a key with a keyring, which means:
* rotate_system_key
* if the redo log initially isn't encrypted, and we request encryption later, we start by using a keyring generated key

Reply:
Ok, regarding the second point, what if that request is startup option?

Am trying to understand the pros and cons here.Can you please list out? 
For MK encryption too, such default key exists but that is changed once uuid is ready.

Reply:
If it is a startup option, and the log isn't encrypted, but it exists? Then we simply generate a key with the keyring. The issue is only when the logs are created during bootstrap, in every other case, the normal way works.

For MK, we *never* change the key. MK generates a redo key at encryption setup, and stores the key in the log file header. This key can't possibly be changed later, as it is needed to read older encrypted records.
But this key isn't stored in plain text, as that would make no sense: MK encrypts the redo key using the innodb master key in the log file header. Every time the master key is rotated, MK reencrypts the log file header. And if MK encryption is turned on before an uuid exists, it encrypts the file header using the default key first (tbh I'm not sure why, because it would never be able to read that header, it doesn't make any sense to write it at all).

Reply:
Thanks, now it is clear what MK does. so let me be specific

After using a default generated key, and assume no rotation initiated, when will redo use a  "percona_redo" key (created by my_key_generate())

Reply:
By "default generated key", you mean a key generated with this without_storing method? If yes, then that key will be saved as percona_redo in the fix_empty_uuid method, so it *is* a percona_redo key. Unless the user rotates it, it won't be changed.

Reply:
ok, checked the code redo_rotate_default_key(). Here we are storing the default key as percona_redo key.

So we cannot generate and new encryption key here (in fix_empty_uuid) and use it for redo tablespace (store it in fil_space_t) and write the key encryption in redo header?

Reply:
This "default" key is as random as any other key, not something fixed, why would we generate a new key?

With keyring encryption, every key we used to encrypt blocks in the redo log has to be present in the keyring. Otherwise we won't be able to decrypt those blocks. Since it's possible that we encrypted blocks with the server generated key, we have to store it.

After that, we could rotate the key - but I don't see the point of it.

Reply:
ok, got it. Here it doesn't make sense (yet). Incase of MK, we have UUID in tablespace encryption header. So in MK, the rotation at fix_empty_uuid,  fixes all tablespaces that wrote tablespace without UUID to headers with UUID.